source/hpack.c

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/http/private/hpack.h>
+
+#include <aws/http/request_response.h>
+
+#include <aws/compression/huffman.h>
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/hash_table.h>
+#include <aws/common/string.h>
+
+#include <assert.h>
+
+struct aws_huffman_symbol_coder *hpack_get_coder(void);
+
+int aws_hpack_encode_integer(uint64_t integer, uint8_t prefix_size, struct aws_byte_buf *output) {
+    assert(prefix_size <= 8);
+
+    if (output->len == output->capacity) {
+        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+    }
+    const uint8_t cut_bits = 8 - prefix_size;
+    const uint8_t prefix_mask = UINT8_MAX >> cut_bits;
+
+    if (integer < prefix_mask) {
+        /* If the integer fits inside the specified number of bits but
+           won't be all 1's, just write it */
+
+        /* Just write out the bits we care about */
+        *output->buffer |= integer;
+        ++output->len;
+    } else {
+        /* Set all of the bits in the first octet to 1 */
+        *output->buffer |= prefix_mask;
+        ++output->len;
+
+        integer -= prefix_mask;
+
+        const uint64_t hi_57bit_mask = UINT64_MAX - (UINT8_MAX >> 1);
+
+        while (integer) {
+            if (output->len == output->capacity) {
+                return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
```

Comment:
This can fail after writing SOME output

we should probably detect the amount of space we need and fail early
Or take some other approach where we use state to track what's partially encoded

Or a failure here i fatal to the user, which is fine if that's the way we expect to use it.